File: sol/solution/0101-0200/0103/Solution.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<IList<int>> ZigzagLevelOrder(TreeNode root) {
        var result = new List<IList<int>>();
        if (root == null) return result;
        var queue = new Queue<TreeNode>();
        queue.Enqueue(root);
        bool leftToRight = true;
        while (queue.Count > 0) {
            int size = queue.Count;
            var level = new List<int>(size);
            for (int i = 0; i < size; i++) {
                var node = queue.Dequeue();
                if (leftToRight) {
                    level.Add(node.val);
                } else {
                    level.Insert(0, node.val);
                }
                if (node.left != null) queue.Enqueue(node.left);
                if (node.right != null) queue.Enqueue(node.right);
            }
            result.Add(level);
            leftToRight = !leftToRight;
        }
        return result;
    }
}

File: sol/solution/0101-0200/0103/Solution.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
  List<List<int>> result = [];
  List<List<int>> zigzagLevelOrder(TreeNode? root) {
    dfs(root, 0);
    return result;
  }
  dfs(TreeNode? node, int level){
    if (node == null) return;
    if (result.length <= level) result.add([]);
    if (level % 2 == 0){
      result[level].add(node.val);
    }else {
      result[level].insert(0, node.val);
    }
    dfs(node.left, level+1);
    dfs(node.right, level+1);

  }
}